pkg/cloud/services/network/eips.go

```diff
@@ -156,3 +185,75 @@ func (s *Service) getEIPTagParams(role string) infrav1.BuildParams {
 		Additional:  s.scope.AdditionalTags(),
 	}
 }
+
+// GetOrAllocateAddresses exports the interface to allocate an address from external services.
+func (s *Service) GetOrAllocateAddresses(pool *infrav1.ElasticIPPool, num int, role string) (eips []string, err error) {
+	return s.getOrAllocateAddresses(num, role, pool)
+}
+
+// ReleaseAddressByRole releases EIP addresses filtering by tag CAPA provider role.
+func (s *Service) ReleaseAddressByRole(role string) error {
+	clusterFilter := []*ec2.Filter{filter.EC2.Cluster(s.scope.Name())}
+	clusterFilter = append(clusterFilter, filter.EC2.ProviderRole(role))
+
+	return s.releaseAddressesWithFilter(clusterFilter)
+}
+
+// setByoPublicIpv4 check if the config has Public IPv4 Pool defined, then
+// check if there are IPs available to consume from allocation, otherwise
+// fallback to Amazon pool when explicty failure isn't defined.
+func (s *Service) setByoPublicIpv4(pool *infrav1.ElasticIPPool, alloc *ec2.AllocateAddressInput) error {
+	if pool == nil {
+		return nil
+	}
+	// check if pool has free IP.
+	ok, err := s.publicIpv4PoolHasFreeIPs(pool, 1)
+	if err != nil {
+		record.Warnf(s.scope.InfraCluster(), "FailedAllocateEIP", "Failed to allocate Elastic IP from Public IPv4 pool %q: %w", *pool.PublicIpv4Pool, err)
+		return fmt.Errorf("failed to update Elastic IP: %w", err)
+	}
+
+	// use the custom public ipv4 pool to the Elastic IP allocation.
+	if ok {
+		alloc.PublicIpv4Pool = pool.PublicIpv4Pool
+		return nil
+	}
+
+	// default, don't change allocation config, use Amazon pool.
+	return nil
+}
+
+// publicIpv4PoolHasFreeIPs check if there are N IPs address available in a Public IPv4 Pool.
+func (s *Service) publicIpv4PoolHasFreeIPs(pool *infrav1.ElasticIPPool, want int64) (bool, error) {
```

Comment:
This function checks if the pool has _at least_ a "want" number of free IPs.
So I'd probably change `publicIpv4PoolHasFreeIPs` to `publicIpv4PoolHasAtLeastNFreeIPs`